cpp/src/arrow/array/array_struct_test.cc

```diff
@@ -256,16 +256,6 @@ TEST_F(TestStructBuilder, TestAppendNull) {
   ASSERT_OK(builder_->AppendNull());
   ASSERT_EQ(2, static_cast<int>(builder_->num_fields()));
 
-  ListBuilder* list_vb = checked_cast<ListBuilder*>(builder_->field_builder(0));
```

Comment:
Is there any reason you removed these lines? The test doesn't look union-related.

Reply:
I mentioned this in the PR description. IMHO it makes sense for both the union and struct builders to take responsibility for invoking AppendNull/AppendNulls on their children rather than UnionBuilder do it and StructBuilder not do it. There were several places where the children were being manually appended to and the new way is definitely cleaner. There's some chance that there is third party code that uses StructBuilder but I'm not sure what to do about this -- I presume such users would be responsible enough to write unit tests.